wpilibc/src/main/native/cpp/LEDPattern.cpp

```diff
@@ -0,0 +1,290 @@
+    units::meters_per_second_t velocity, units::meter_t ledSpacing) {
+  // Velocity is in terms of meters per second
+  // Multiply by 1,000,000 to use microseconds instead of seconds
+  auto microsPerLed =
+      static_cast<int64_t>(std::floor((ledSpacing / velocity).value() * 1e6));
+
+  return LEDPattern{[=, this](auto data, auto writer) {
+    auto bufLen = data.size();
+    auto now = wpi::Now();
+
+    // every step in time that's a multiple of microsPerLED will increment
+    // the offset by 1
+    // cast unsigned int64 `now` to a signed int64 so we can get negative
+    // offset values for negative velocities
+    auto offset = static_cast<int64_t>(now) / microsPerLed;
+
+    ApplyTo(data, [=, &writer](int i, Color color) {
+      // floorMod so if the offset is negative, we still get positive outputs
+      int shiftedIndex = floorMod(i + offset, bufLen);
+
+      writer(shiftedIndex, color);
+    });
+  }};
+}
+
+LEDPattern LEDPattern::Blink(units::second_t onTime, units::second_t offTime) {
+  auto totalMicros = units::microsecond_t{onTime + offTime}.to<uint64_t>();
+  auto onMicros = units::microsecond_t{onTime}.to<uint64_t>();
+
+  return LEDPattern{[=, this](auto data, auto writer) {
+    if (wpi::Now() % totalMicros < onMicros) {
+      ApplyTo(data, writer);
+    } else {
+      LEDPattern::kOff.ApplyTo(data, writer);
+    }
+  }};
+}
+
+LEDPattern LEDPattern::Blink(units::second_t onTime) {
+  return LEDPattern::Blink(onTime, onTime);
+}
+
+LEDPattern LEDPattern::SynchronizedBlink(std::function<bool()> signal) {
+  return LEDPattern{[=, this](auto data, auto writer) {
+    if (signal()) {
+      ApplyTo(data, writer);
+    } else {
+      LEDPattern::kOff.ApplyTo(data, writer);
+    }
+  }};
+}
+
+LEDPattern LEDPattern::Breathe(units::second_t period) {
+  auto periodMicros = units::microsecond_t{period};
+
+  return LEDPattern{[periodMicros, this](auto data, auto writer) {
+    ApplyTo(data, [&writer, periodMicros](int i, Color color) {
+      double t = (wpi::Now() % periodMicros.to<uint64_t>()) /
+                 periodMicros.to<double>();
+      double phase = t * 2 * std::numbers::pi;
+
+      // Apply the cosine function and shift its output from [-1, 1] to [0, 1]
+      // Use cosine so the period starts at 100% brightness
+      double dim = (std::cos(phase) + 1) / 2.0;
+
+      writer(i, Color{color.red * dim, color.green * dim, color.blue * dim});
+    });
+  }};
+}
+
+LEDPattern LEDPattern::OverlayOn(LEDPattern& base) {
+  return LEDPattern{[this, &base](auto data, auto writer) {
+    // write the base pattern down first...
+    base.ApplyTo(data, writer);
+
+    // ... then, overwrite with illuminated LEDs from the overlay
+    ApplyTo(data, [&](int i, Color color) {
+      if (color.red > 0 || color.green > 0 || color.blue > 0) {
+        writer(i, color);
+      }
+    });
+  }};
+}
+
+LEDPattern LEDPattern::Blend(LEDPattern& other) {
+  return LEDPattern{[&](auto data, auto writer) {
```

Comment:
These & captures are not safe, as they capture references to objects (both `this` and `other`) which can be moved or deleted after this function returns.  What you will need to do here is *copy* the objects (including `this`), or copy their implementation functions, e.g. `[self = *this, other]`.